docs/fheroes2.6
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2022  fheroes2 team <fhomm2@gmail.com>
.\"
.TH fheroes2 6 "May 2022"
.\" Please adjust this date whenever revising the manpage.
.SH NAME
fheroes2 \- free remake of Heroes of Might and Magic II game engine
.SH SYNOPSIS
.B fheroes2
.SH DESCRIPTION
\fBfheroes2\fR is a free implementation of Heroes of Might and Magic II game engine,
a classic turn-based strategy, with significant improvements in the gameplay, graphics
and logic (including support for high-resolution graphics, improved AI,
numerous fixes and UI improvements).
.PP
To play the game, original resources from a demo or full version are needed.
.SH GAME DATA PATHS 
The following directories are searched as data root:
.TP
.B /usr/share/games/fheroes2
The system-wide location that can be accessed by all users.
.TP
.B FHEROES2_DATA
This environment variable can contain a path to a directory with game files.
.TP
.B Current working directory
.TP
.B $HOME/.local/share/fheroes2
Per-user data directory. This path can be overridden by the XDG_DATA_HOME environment variable.
.PP
The game resources which searched in these directories:
.TP
.B data/*.agg
Original assets (required).
.TP
.B files/data/*.h2d
Additional non-original assets from fheroes2 project.
.TP
.B files/lang/*.mo
Translations from fheroes2 project.
.TP
.B maps/*
.br
Single scenarios and campaigns (h2c, hxc, mp2, mx2 files).
.TP
.B music/*
CD audio music tracks which are ripped in mp3, ogg or flac formats.
The tracks should be named as "homm2_XX.ext" or "TrackXX.ext" where XX is a track number and ext is an extension.
Numeration starts from "01" for the first naming and from "02" for the second one.
.TP
\fBanim/*.smk\fP or \fBanim2/*.smk\fP or \fBheroes2/anim/*.smk\fP
Video files.
.SH ENVIRONMENT
.TP
.B FHEROES2_DATA
See the section \fBGAME DATA PATHS\fR above.
.SH FILES
.TP
.B $HOME/.config/fheroes2
This directory contains per-user configuration files.
\fBfheroes2.cfg\fR is for game options and \fBfheroes2.key\fR is for hotkeys.
This path can be overridden by the XDG_CONFIG_HOME environment variable.



changelog.txt
version 0.9.15 (07 May 2022)
- add HotKey window and expand the list of system options
- do not allow higher-level Mage Guilds to be built if the previous-level Mage Guild has not been built yet
- Battle Only mode: always assign a random spell to the Spell Scroll
- add initial support for Romanian language
- add more conditions to define a blocking AI hero
- fix incorrect video playback on Descendants campaign, scenario 5
- change some descriptions of campaign awards and bonuses
- do not auto save after the completion of the last scenario in a campaign
- add images and full description of bonuses and awards by right mouse click
- fix some incorrect campaign bonuses
- display None if no awards available for the current scenario
- add hot keys to select bonuses
- fix rendering of campaign UI window as it was very slow on response on changing bonuses
- fix missing kingdom income value for scenarios with additional bonus resources
- game type interface for campaign window must be unique and should not depend on overall game interface
- allow already selected scenario to appear in ALL scenarios list
- improved Week of the creature window
- changed town portal's text and condition to be similar to town gate
- update Russian translation
- properly extract animations from the GOG distribution of HoMM2 for Windows
- update French translation
- improve search for video files
- add text truncation method to be used in campaign's choices and awards
- encourage AI to explore more
- redesign Hot Key event file handling
- add double-lined headers to adventure map options
- add Belarusian, Bulgarian and Ukrainian languages support
- populate resolution list based till the lowest resolution
- fix 1000 and 100000 short values display
- update hero quick info popup dialog
- disable Town Portal dialog OK button by default
- changed AI's turn to be repeated multiple times for heroes
- generate French 'MIN' button
- generate French 'GIFT' button
- add Translation Context for Secondary Skill Campaign Bonuses
- add Ogre Alliance and Dwarfbane full strings for translation
- fix the positioning of the damage info popup
- update Norwegian translation
- fix calculation of morale & luck modifiers for in-castle army without a commander
- properly calculate the number of travel days to the destination tile for a hero
- AI should replenish lost spell points more frequently
- do not allow a hero to have multiple spell books in the "Battle Only" mode
- fix rendering of the Arm of the Martyr's shadow
- AI uses View All spell to scout fog
- hero should have at least 69MP to cast Dimension Door, Town Gate and Town Portal spells
- fix spacing of player squares in Select Recipients window and properly center resource icons for "Your Funds" and "Planned Gift" horizontally
- fix French small font Cedilla c rendering
- update Polish translation
- reserve half of AI spell points for Dimension Door spell
- hero should be able to cast spells that consume movement points if he is able to move from his current tile
- fix MP3 support in Windows SDL1 packages
- reset the mixer volume before playing video soundtracks
- expand Test Support mode to cover all campaign related actions
- add Black-White cursor support
- add music control options to the configuration dialog
- fix failing music file search
- properly work with guardians on the objects that can be captured
- fix call to get secondary skill name leading to a crash
- fix multiple issues related to the interface settings
- allow to change the auto battle settings if player's army doesn't have a commander
- fix a crash when a hero tries to get a second magic book
- always grab the mouse in fullscreen mode to properly support devices with notch

version 0.9.14 (09 April 2022)
- generate Italian "Battle Only" button
- fix line breaking in welcome message for various languages
- update and improve Italian translation
- generate Polish "Battle Only" button
- fix switching fullscreen mode in Mac OS for certain resolutions
- generate French "Battle Only" button
- fix a crash due to an assertion failure when the shooter hits back at a unit under Berserk spell
- fix scrolling in full-screen mode in case of an aspect ratio mismatch
- fix rare cases for incorrect letters being generated while switching languages
- add Norwegian language option and font generation
- update French translation
- add initial implementation of Dimension Door spell usage for AI
- fix Gold Watch artifact invalid bonus
- rework artifact logic with proper separation and calculations of artifact's bonuses and curses
- play only one sound while getting ar artifact within an event
- speed up castle and hero icons on adventure map
- add "Battle Only" button generation for German version of the game
- fix Mage Guild bottom status bar rendering
- fix incorrect artifact discovery description
- add in-game Italian language option and diacritics
- fix army order option related crash
- fix High Score Music after winning scenario
- adjust AI spell estimates
- fix rendering of bottom bar in Well dialog
- update Polish translation
- remove unnecessary call to play VICTORY music during Campaign intro
- fix the calculation of the cost of surrender
- fix Town Gate spell logic
- prioritize AI castles further from enemy territory
- take the weekly growth into account when calculating monthly population growth bonuses and fines
- update minimap colors
- optimize AI fighter hero priorities
- add native macOS application generation
- build player influence map to inform AI kingdom development decisions
- speed up data reading from AGG files
- update Norwegian translation
- add a popup window for battle log
- add a mandatory requirement of h2d file being present 
- fix incorrect logic for palette application function
- fix position of armies for quick info window for both castle and enemy hero
- set the direction for attacking and defending units in battle in a more OG-like way
- fix incorrect popup monster window in Well

src/fheroes2/game/fheroes2.cpp
/***************************************************************************
 *   fheroes2: https://github.com/ihhub/fheroes2                           *
 *   Copyright (C) 2019 - 2022                                             *
 *                                                                         *
 *   Free Heroes2 Engine: http://sourceforge.net/projects/fheroes2         *
 *   Copyright (C) 2009 by Andrey Afletdinov <fheroes2@gmail.com>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <cstdlib>
#include <iostream>
#include <string>

#include "agg.h"
#include "audio_manager.h"
#include "bin_info.h"
#include "core.h"
#include "cursor.h"
#include "embedded_image.h"
#include "game.h"
#include "game_logo.h"
#include "game_video.h"
#include "h2d.h"
#include "image_palette.h"
#include "localevent.h"
#include "logging.h"
#include "screen.h"
#include "settings.h"
#include "system.h"
#ifdef WITH_DEBUG
#include "tools.h"
#endif
#include "ui_tool.h"
#include "zzlib.h"

namespace
{
    std::string GetCaption()
    {
        return std::string( "fheroes2 engine, version: " + Settings::GetVersion() );
    }

    int PrintHelp( const char * basename )
    {
        COUT( "Usage: " << basename << " [OPTIONS]" )
#ifdef WITH_DEBUG
        COUT( "  -d <level>\tprint debug messages, see src/engine/logging.h for possible values of <level> argument" )
#endif
        COUT( "  -h\t\tprint this help message and exit" )

        return EXIT_SUCCESS;
    }

    void ReadConfigs()
    {
        const std::string configurationFileName( Settings::configFileName );
        const std::string confFile = Settings::GetLastFile( "", configurationFileName );

        Settings & conf = Settings::Get();
        if ( System::IsFile( confFile ) && conf.Read( confFile ) ) {
            LocalEvent::Get().SetControllerPointerSpeed( conf.controllerPointerSpeed() );
        }
        else {
            conf.Save( configurationFileName );
        }
    }

    void InitConfigDir()
    {
        const std::string configDir = System::GetConfigDirectory( "fheroes2" );

        if ( !System::IsDirectory( configDir ) ) {
            System::MakeDirectory( configDir );
        }
    }

    void InitDataDir()
    {
        const std::string dataDir = System::GetDataDirectory( "fheroes2" );

        if ( dataDir.empty() )
            return;

        const std::string dataFiles = System::ConcatePath( dataDir, "files" );
        const std::string dataFilesSave = System::ConcatePath( dataFiles, "save" );

        if ( !System::IsDirectory( dataDir ) )
            System::MakeDirectory( dataDir );

        if ( System::IsDirectory( dataDir, true ) && !System::IsDirectory( dataFiles ) )
            System::MakeDirectory( dataFiles );

        if ( System::IsDirectory( dataFiles, true ) && !System::IsDirectory( dataFilesSave ) )
            System::MakeDirectory( dataFilesSave );
    }

    class DisplayInitializer
    {
    public:
        DisplayInitializer()
        {
            const Settings & conf = Settings::Get();

            fheroes2::engine().setVSync( conf.isVSyncEnabled() );

            fheroes2::Display & display = fheroes2::Display::instance();
            if ( conf.FullScreen() != fheroes2::engine().isFullScreen() )
                fheroes2::engine().toggleFullScreen();

            display.resize( conf.VideoMode().width, conf.VideoMode().height );
            display.fill( 0 ); // start from a black screen

            fheroes2::engine().setTitle( GetCaption() );

            SDL_ShowCursor( SDL_DISABLE ); // hide system cursor

            // Initialize local event processing.
            LocalEvent::Get().RegisterCycling( fheroes2::PreRenderSystemInfo, fheroes2::PostRenderSystemInfo );

            // Update mouse cursor when switching between software emulation and OS mouse modes.
            fheroes2::cursor().registerUpdater( Cursor::Refresh );

#if !defined( MACOS_APP_BUNDLE )
            const fheroes2::Image & appIcon = CreateImageFromZlib( 32, 32, iconImage, sizeof( iconImage ), true );
            fheroes2::engine().setIcon( appIcon );
#endif
        }

        DisplayInitializer( const DisplayInitializer & ) = delete;
        DisplayInitializer & operator=( const DisplayInitializer & ) = delete;

        ~DisplayInitializer()
        {
            fheroes2::Display::instance().release();
        }
    };

    class DataInitializer
    {
    public:
        DataInitializer()
        {
            const fheroes2::ScreenPaletteRestorer screenRestorer;

            try {
                _aggInitializer.reset( new AGG::AGGInitializer );

                _h2dInitializer.reset( new fheroes2::h2d::H2DInitializer );
            }
            catch ( ... ) {
                fheroes2::Display & display = fheroes2::Display::instance();
                const fheroes2::Image & image = CreateImageFromZlib( 290, 190, errorMessage, sizeof( errorMessage ), false );

                display.fill( 0 );
                fheroes2::Resize( image, display );

                display.render();

                LocalEvent & le = LocalEvent::Get();
                while ( le.HandleEvents() && !le.KeyPress() && !le.MouseClickLeft() ) {
                    // Do nothing.
                }

                throw;
            }
        }

        DataInitializer( const DataInitializer & ) = delete;
        DataInitializer & operator=( const DataInitializer & ) = delete;
        ~DataInitializer() = default;

        const std::string & getOriginalAGGFilePath() const
        {
            return _aggInitializer->getOriginalAGGFilePath();
        }

        const std::string & getExpansionAGGFilePath() const
        {
            return _aggInitializer->getExpansionAGGFilePath();
        }

    private:
        std::unique_ptr<AGG::AGGInitializer> _aggInitializer;
        std::unique_ptr<fheroes2::h2d::H2DInitializer> _h2dInitializer;
    };
}

#if defined( _MSC_VER )
#undef main
#endif

int main( int argc, char ** argv )
{
    try {
        const fheroes2::HardwareInitializer hardwareInitializer;
        Logging::InitLog();

        COUT( GetCaption() )

        Settings & conf = Settings::Get();
        conf.SetProgramPath( argv[0] );

        InitConfigDir();
        InitDataDir();
        ReadConfigs();

        // getopt
        {
            int opt;
            while ( ( opt = System::GetCommandOptions( argc, argv, "hd:" ) ) != -1 )
                switch ( opt ) {
#ifdef WITH_DEBUG
                case 'd':
                    conf.SetDebug( System::GetOptionsArgument() ? GetInt( System::GetOptionsArgument() ) : 0 );
                    break;
#endif
                case '?':
                case 'h':
                    return PrintHelp( argv[0] );

                default:
                    break;
                }
        }

        std::set<fheroes2::SystemInitializationComponent> coreComponents{ fheroes2::SystemInitializationComponent::Audio,
                                                                          fheroes2::SystemInitializationComponent::Video };

#if defined( TARGET_PS_VITA ) || defined( TARGET_NINTENDO_SWITCH )
        coreComponents.emplace( fheroes2::SystemInitializationComponent::GameController );
#endif

        const fheroes2::CoreInitializer coreInitializer( coreComponents );

        DEBUG_LOG( DBG_GAME, DBG_INFO, conf.String() )

        const DisplayInitializer displayInitializer;

        const DataInitializer dataInitializer;

        const AudioManager::AudioInitializer audioInitializer( dataInitializer.getOriginalAGGFilePath(), dataInitializer.getExpansionAGGFilePath() );

        // Load palette.
        fheroes2::setGamePalette( AGG::getDataFromAggFile( "KB.PAL" ) );
        fheroes2::Display::instance().changePalette( nullptr, true );

        // load BIN data
        Bin_Info::InitBinInfo();

        // init game data
        Game::Init();

        conf.setGameLanguage( conf.getGameLanguage() );

        if ( conf.isShowIntro() ) {
            fheroes2::showTeamInfo();

            Video::ShowVideo( "H2XINTRO.SMK", Video::VideoAction::PLAY_TILL_VIDEO_END );
        }

        // init cursor
        const CursorRestorer cursorRestorer( true, Cursor::POINTER );

        Game::mainGameLoop( conf.isFirstGameRun() );
    }
    catch ( const std::exception & ex ) {
        ERROR_LOG( "Exception '" << ex.what() << "' occurred during application runtime." )
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
